core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -614,18 +614,24 @@ private[spark] class ExecutorAllocationManager(
             stageAttemptToNumRunningTask -= stageAttempt
           }
         }
-        // If the task failed, we expect it to be resubmitted later. To ensure we have
-        // enough resources to run the resubmitted task, we need to mark the scheduler
-        // as backlogged again if it's not already marked as such (SPARK-8366)
-        if (taskEnd.reason != Success) {
-          if (totalPendingTasks() == 0) {
-            allocationManager.onSchedulerBacklogged()
-          }
-          if (taskEnd.taskInfo.speculative) {
-            stageAttemptToSpeculativeTaskIndices.get(stageAttempt).foreach {_.remove(taskIndex)}
-          } else {
-            stageAttemptToTaskIndices.get(stageAttempt).foreach {_.remove(taskIndex)}
-          }
+
+        if (taskEnd.taskInfo.speculative) {
+          stageAttemptToSpeculativeTaskIndices.get(stageAttempt).foreach {_.remove{taskIndex}}
+          stageAttemptToNumSpeculativeTasks(stageAttempt) -= 1
```

Comment:
one question I actually have here is what happens on a speculative task failure.  The scheduler will try to rerun but we may not be accounting for it if we have removed, I need to look at the scheduler logic a bit more.

Reply:
When a speculative task fails while the original task is running, it will not be resubmitted (because we have one copy running, see https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L289-L293), but a new speculative task will be launched later in a future speculation check cycle. The code works correctly in this case, as we updated both `stageAttemptToSpeculativeTaskIndices` and `stageAttemptToNumSpeculativeTasks`, so this failed speculative task will just behave as it never existed. I can add this to my test case.

@tgravescs, @jiangxb1987: Note that the current design (with or without this PR) will launch more-than-needed executors when **the original task fails when the speculative task is running**. In this case, the original task will not be resubmitted (same as above because we have one copy running). Speculation check will later launch one more speculative task on the already running speculative task. In this case, **we have 2 speculative tasks for the same task index**. In this case, we will still launch more than needed executors.

Reply:
thanks for the explanation, its been a while since I looked at the speculative handling in the scheduler.   I think the failed task case with speculative is ok. I think the problem is the scheduler isn't  tracking those task as speculative or not after the initial launch.  I would rather error on the side of asking for to many and the failure case should be less frequent.  that is a lot more complex case along with making the dynamic allocation manager smarter about it.  If you can file that other jira I mentioned above to follow up, if you could put those details in there as well that would be great.

Reply:
sorry I don't follow the logic. According to `TSM.dequeueTask()`:
>     // Tries to schedule a regular task first; if it returns None, then schedules
>     // a speculative task
It would first try to schedule a regular task (which would return None in this case), then it would try to schedule a speculative task. Since we have only the original task running, it shall schedule one speculative task in this case. Am I missing something? I also searched the test case but didn't locate the place describing the behavior.

Reply:
> sorry I don't follow the logic. According to `TSM.dequeueTask()`:
> 
> > ```
> > // Tries to schedule a regular task first; if it returns None, then schedules
> > // a speculative task
> > ```
> 
> It would first try to schedule a regular task (which would return None in this case), then it would try to schedule a speculative task. Since we have only the original task running, it shall schedule one speculative task in this case. Am I missing something? I also searched the test case but didn't locate the place describing the behavior.

@jiangxb1987 Let's make sure we are on the same page. I assume you are talking about the case on `speculation task failure`. As I have explained above, when a speculative task fails, `handleFailedTask` will be invoked, and we will have `addPendingTask(index)` at L894. This will add a pending task for the failed task index (with `speculatable` default to false).

Then from `TSM.dequeueTask()`, we have a call chain of `dequeueTaskHelper` -> `dequeueTaskFromList`, see https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L289-L293: as we already have one copy running (the original task), the enqueued pending task will not be resubmitted here. Instead, in a future speculation cycle, we will see the original task running too slow and will try to speculate again on the original task from there.

Reply:
I guess you might be questioning on the code itself. As I have explained above, a failed speculative task will not be resubmitted, but instead a new speculative task will be launched in a future speculation cycle.

Reply:
Yes, there seems to be a bug at L894, a failed speculative task should be add to the pendingSpeculatableTasks. But as you explained, your fix is correct here.